### PR TITLE
Fix Forgotten Knowledge Bosses

### LINKED ISSUE #285: BUG: Dragon Egg Bug
#### Describe the wrong or missing content

Hello, when it comes to killing the boss of the ice portal in the room where you need to protect the dragon egg to become a baby dragon, you are giving mistakes 

#### Describe the correct content

as verified in the quest, it needs to turn the monster into another one and after it kills, the egg needs to hatch

#### Screenshots
unknown

### LINKED ISSUE #363: [Bug]: Energy portal issue, Lloyd
### What happened?

So, you can attack and kill the wrong Energy while Lloyd is regenerating.
He's alive:
https://i.gyazo.com/6a7c8ef77220b0012eb49d35a3c00bed.png
He's regen..
https://i.gyazo.com/df5207f578e7ac300dfae64af383b845.png

So, it's not healing itself if Lloyd isn't "alive"

### How to Reproduce?

X: 32799 Y: 32831 Z: 14

### Relevant log output

_No response_

### Attach any relevant file

https://i.gyazo.com/3017d74c0ef5e7deac97e637978d0358.png
That's the only error i get after slaying him, an error/warning of an item conversion 

### Version

12.x (Default)

### What OS are you seeing the problem on?

Ubuntu 20.04

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by murilo09**
# Description

Fix Lloyd fight and Forgotten Knowledge bosses time

Remove AID from positions: 
- [x] 33391, 31177, 10
- [x] 32019, 32843, 14
- [x] 32302, 31087, 14
- [x] 32657, 32876, 14
- [x] 32759, 32867, 14
- [x] 32902, 31622, 14
- [x] 33010, 31659, 14

### **Actual**

Can kill any prism in Lloyd fight

### **Expected**

Now Lloyd is working as expected

## Fixes

#285 
#363 

## Type of change

Please delete options that are not relevant.

  - [x] New feature (non-breaking change which adds functionality)

## Checklist

  - [X] My code follows the style guidelines of this project
  - [X] I have performed a self-review of my own code
  - [X] I checked the PR checks reports
  - [X] My changes generate no new warnings

### CODE REVIEW

**Comment by murilo09** [comment]
Drafted while the other bosses are not done

**Review by majestyotbr — APPROVED** [review]

File: data/scripts/creaturescripts/quests/forgotten_knowledge/melting_death.lua [hunk 1]
```diff
@@ -0,0 +1,16 @@
+local monster = {
+    name = 'baby dragon',
+    pos = Position(32269, 31084, 14)
+}
+
+local meltingDeath = CreatureEvent("MeltingDeath")
+function meltingDeath.onDeath(creature, corpse, lasthitkiller, mostdamagekiller, lasthitunjustified, mostdamageunjustified)
+    if creature:isMonster() and creature:getName():lower() ~= 'melting frozen horror' then
```
Comment (dudantas) [inline data/scripts/creaturescripts/quests/forgotten_knowledge/melting_death.lua]:
```suggestion
    if not creature then
        return
    end
   
    if creature:isMonster() and creature:getName():lower() ~= 'melting frozen horror' then
```

File: data/scripts/actions/quests/forgotten_knowledge/dragonking_lever.lua [hunk 2]
```diff
@@ -1,61 +1,105 @@
...
+                end
+            end
+            spec:removePlayers(player_remove)
+        end, config.timeToDefeatBoss * 60 * 1000)
+    end
 end
 
-forgottenKnowledgeDragonking:aid(24880)
+forgottenKnowledgeDragonking:position(Position(33391, 31177, 10))
 forgottenKnowledgeDragonking:register()
```
Comment (dudantas) [inline data/scripts/actions/quests/forgotten_knowledge/dragonking_lever.lua]:
```suggestion
forgottenKnowledgeDragonking:register()

```

File: data/scripts/actions/quests/forgotten_knowledge/horror_lever.lua [hunk 3]
```diff
@@ -1,58 +1,103 @@
...
+                end
+            end
+            spec:removePlayers(player_remove)
+        end, config.timeToDefeatBoss * 60 * 1000)
+    end
 end
 
-forgottenKnowledgeHorror:aid(24882)
+forgottenKnowledgeHorror:position(Position(32302, 31087, 14))
 forgottenKnowledgeHorror:register()
```
Comment (dudantas) [inline data/scripts/actions/quests/forgotten_knowledge/horror_lever.lua]:
```suggestion
forgottenKnowledgeHorror:register()

```

File: data/scripts/actions/quests/forgotten_knowledge/tenebris_lever.lua [hunk 4]
```diff
@@ -1,65 +1,96 @@
...
+                end
+            end
+            spec:removePlayers(player_remove)
+        end, config.timeToDefeatBoss * 60 * 1000)
+    end
 end
 
-forgottenKnowledgeTenebris:aid(24878)
+forgottenKnowledgeTenebris:position(Position(32902, 31622, 14))
 forgottenKnowledgeTenebris:register()
```
Comment (dudantas) [inline data/scripts/actions/quests/forgotten_knowledge/tenebris_lever.lua]:
```suggestion
forgottenKnowledgeTenebris:register()

```

File: data/scripts/actions/quests/forgotten_knowledge/thorn_knight_lever.lua [hunk 5]
```diff
@@ -1,51 +1,96 @@
...
+                end
+            end
+            spec:removePlayers(player_remove)
+        end, config.timeToDefeatBoss * 60 * 1000)
+    end
 end
 
-forgottenKnowledgeThorn:aid(24879)
+forgottenKnowledgeThorn:position(Position(32657, 32876, 14))
 forgottenKnowledgeThorn:register()
```
Comment (dudantas) [inline data/scripts/actions/quests/forgotten_knowledge/thorn_knight_lever.lua]:
```suggestion
forgottenKnowledgeThorn:register()

```

File: data/scripts/actions/quests/forgotten_knowledge/time_guardian_lever.lua [hunk 6]
```diff
@@ -1,55 +1,106 @@
...
+                end
+            end
+            spec:removePlayers(player_remove)
+        end, config.timeToDefeatBoss * 60 * 1000)
+    end
 end
 
-forgottenKnowledgeGuardianLever:aid(24883)
+forgottenKnowledgeGuardianLever:position(Position(33010, 31659, 14))
 forgottenKnowledgeGuardianLever:register()
```
Comment (dudantas) [inline data/scripts/actions/quests/forgotten_knowledge/time_guardian_lever.lua]:
```suggestion
forgottenKnowledgeGuardianLever:register()

```

File: data/scripts/creaturescripts/quests/forgotten_knowledge/energy_prism_death.lua [hunk 7]
```diff
@@ -1,26 +1,11 @@
...
 	local lloyd = Tile(Position(32799, 32826, 14)):getTopCreature()
 	if lloyd then
-		lloyd:teleportTo(Position(32799, 32828, 14))
+		lloyd:teleportTo(Position(32799, 32829, 14))
 		lloyd:getPosition():sendMagicEffect(CONST_ME_TELEPORT)
 	end
 	return true
 end
-energyPrismDeath:register()
+energyPrismDeath:register()
```
Comment (dudantas) [inline data/scripts/creaturescripts/quests/forgotten_knowledge/energy_prism_death.lua]:
```suggestion
energyPrismDeath:register()

```

File: data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua [hunk 8]
```diff
@@ -1,48 +1,50 @@
...
+    Game.createMonster(monsters[prismId].cosmicInvu, Position(monsters[prismId].pos), true, true)
 end
 
 local lloydPrepareDeath = CreatureEvent("LloydPrepareDeath")
 function lloydPrepareDeath.onPrepareDeath(creature, lastHitKiller, mostDamageKiller)
-	if not creature:getName():lower() == "lloyd" and creature:isMonster() then
-		return true
-	end
+    local targetMonster = creature:getMonster()
+    if not targetMonster then
```
Comment (dudantas) [inline data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua]:
```suggestion
    if not creature or not targetMonster then
```

**Review by dudantas — CHANGES_REQUESTED** [review]

File: data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua [hunk 9]
```diff
@@ -1,48 +1,50 @@
...
-		monster:say('The cosmic energies in the chamber refocus on Lloyd.', TALKTYPE_MONSTER_SAY)
-		monster:teleportTo(Position(32799, 32828, 14))
-		monster:getPosition():sendMagicEffect(CONST_ME_TELEPORT)
-	end
+local function revertLloyd(prismId)
+    local lloyd = Tile(Position(32799, 32826, 14)):getTopCreature()
+	if lloyd then
+		lloyd:teleportTo(Position(32799, 32829, 14))
+		lloyd:getPosition():sendMagicEffect(CONST_ME_TELEPORT)
+    end
```
Comment (dudantas) [inline data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua]:
```suggestion
local function revertLloyd(prismId)
	local lloydTile = Tile(Position(32799, 32826, 14))
	if lloydTile then
		local lloyd = lloydTile:getTopCreature()
		lloyd:teleportTo(Position(32799, 32829, 14))
		lloyd:getPosition():sendMagicEffect(CONST_ME_TELEPORT)
	end
```

File: data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua [hunk 10]
```diff
@@ -1,48 +1,50 @@
 local monsters = {
-	{cosmic = 'cosmic energy prism a', pos = Position(32801, 32827, 14)},
-	{cosmic = 'cosmic energy prism b', pos = Position(32798, 32827, 14)},
-	{cosmic = 'cosmic energy prism c', pos = Position(32803, 32826, 14)},
-	{cosmic = 'cosmic energy prism d', pos = Position(32796, 32826, 14)}
+    {cosmicNormal = 'cosmic energy prism a', cosmicInvu = 'cosmic energy prism a invu', pos = Position(32801, 32827, 14)},
+    {cosmicNormal = 'cosmic energy prism b', cosmicInvu = 'cosmic energy prism b invu', pos = Position(32798, 32827, 14)},
+    {cosmicNormal = 'cosmic energy prism c', cosmicInvu = 'cosmic energy prism c invu', pos = Position(32803, 32826, 14)},
+    {cosmicNormal = 'cosmic energy prism d', cosmicInvu = 'cosmic energy prism d invu', pos = Position(32796, 32826, 14)}
```
Comment (dudantas) [inline data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua]:
```suggestion
	{cosmicNormal = 'cosmic energy prism a', cosmicInvu = 'cosmic energy prism a invu', pos = Position(32801, 32827, 14)},
	{cosmicNormal = 'cosmic energy prism b', cosmicInvu = 'cosmic energy prism b invu', pos = Position(32798, 32827, 14)},
	{cosmicNormal = 'cosmic energy prism c', cosmicInvu = 'cosmic energy prism c invu', pos = Position(32803, 32826, 14)},
	{cosmicNormal = 'cosmic energy prism d', cosmicInvu = 'cosmic energy prism d invu', pos = Position(32796, 32826, 14)}
```

File: data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua [hunk 11]
```diff
@@ -1,48 +1,50 @@
...
-		local cosmics = monsters[i]
-		local cosmic = Tile(Position(cosmics.pos)):getTopCreature()
-		if cosmic and cosmic:isMonster() and cosmic:getName():lower() == cosmics.cosmic and cosmic:getHealth() >= 1 then
-			cosmic:setMaxHealth(100000)
-			cosmic:addHealth(100000)
-			return true
-		end
-	end
+    Tile(monsters[prismId].pos):getTopCreature():remove()
+    Game.createMonster(monsters[prismId].cosmicInvu, Position(monsters[prismId].pos), true, true)
```
Comment (dudantas) [inline data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua]:
```suggestion
	Tile(monsters[prismId].pos):getTopCreature():remove()
	Game.createMonster(monsters[prismId].cosmicInvu, Position(monsters[prismId].pos), true, true)
```

File: data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua [hunk 12]
```diff
@@ -1,48 +1,50 @@
...
 
 local lloydPrepareDeath = CreatureEvent("LloydPrepareDeath")
 function lloydPrepareDeath.onPrepareDeath(creature, lastHitKiller, mostDamageKiller)
-	if not creature:getName():lower() == "lloyd" and creature:isMonster() then
-		return true
-	end
+    local targetMonster = creature:getMonster()
+    if not creature or not targetMonster then
+        return true
+    end
```
Comment (dudantas) [inline data/scripts/creaturescripts/quests/forgotten_knowledge/lloyd_preparedeath.lua]:
```suggestion
	local targetMonster = creature:getMonster()
	if not creature or not targetMonster then
		return true
	end
```

File: data/scripts/actions/quests/forgotten_knowledge/dragonking_lever.lua [hunk 13]
```diff
@@ -1,61 +1,105 @@
...
+	spec:check()
+
+    if spec:getPlayers() > 0 then
+        player:sendTextMessage(MESSAGE_EVENT_ADVANCE, "There's someone fighting with " .. config.bossName .. ".")
+        return true
+    end
+
+    local lever = Lever()
+    lever:setPositions(config.playerPositions)
+    lever:setCondition(function(creature)
```
Comment (dudantas) [inline data/scripts/actions/quests/forgotten_knowledge/dragonking_lever.lua]: Where does this creature come from?

**Review by dudantas — APPROVED** [review]

**Pull request merged**